### PR TITLE
Bug 1835371: router/metrics: Fix haproxy_server_max_sessions test

### PR DESCRIPTION
**Pull request by frobware**
This assertion currently flakes a lot. (I believe) the flake occurs
when the router reloads and the GET for the latest metrics reads
zero's because there has been no updates.

It may be better to mark this test as serial. This commit retries the
assertion as opposed to dealing with it as a one-shot.

The expectation is that haxproxy_server_max_sessions needs to be
greater than zero over two successive scrapes once the first scrape
shows an increase. However, this assertion can still fail depending on
when the metrics are refreshed and when the router reloads, which is
every 5s. In my testing I have seen this to be more resilient but
there are periods when running the full openshift/conformance tests
that this continues to fail.

### CODE REVIEW

**Comment by frobware** [comment]
/hold

It's not clear this much better as in my testing this still flakes.

**Comment by frobware** [comment]
Under load and with other tests running I'm seeing it take:

    May 29 15:12:09.594: INFO: metrics refresh completed in 6s

to get new metric data. As the router is reloading every 5s it's possible that we read zeros as and when haproxy restarts.

**Comment by frobware** [comment]
@smarterclayton not sure whether I'm heading down the right path here. It's better, but doesn't eliminate the flakes.

**Comment by frobware** [comment]
I just started a local run for 100 iterations of the same test together with the openshift/conformance suite running against the same cluster. Will report on the pass/fail rate using this approach.

**Comment by frobware** [comment]
Using this approach I do see that we get to:

    scrape 0

then a lot of times that does move on to:

    scrape 0
    scrape 1

and does get to:

    scrape 2 (which is success)

but more often that not we eventually timeout, with debug/info show getting to scrape 1.

**Comment by frobware** [comment]
No haproxy metrics failure in this run:

https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-gcp/8122

/retest

**Comment by frobware** [comment]
Running this test repeatedly I currently have 7 failures in 70 invocations. The conformance tests have been running throughout those invocations.

File: test/extended/router/metrics.go [hunk 1]
```diff
@@ -186,7 +186,73 @@ var _ = g.Describe("[sig-network][Feature:Router]", func() {
...
+					if err != nil {
+						return nil, err
+					}
+					metrics, err := p.TextToMetricFamilies(bytes.NewBufferString(results))
+					e2e.Logf("metrics refresh completed in %s", time.Now().Sub(startTime).Round(time.Second).String())
+					return metrics, err
+				}
+
+				maxSessionsInvariant := func(v []float64) bool {
+					return len(v) >= 2 && v[0] > 0 && v[1] > 0
```
Comment (sgreene570) [inline test/extended/router/metrics.go:207]: Are there situations in this test where the `maxSessionsInvariant` function should return true for slices with more than 2 values?

File: test/extended/router/metrics.go [hunk 2]
```diff
@@ -186,7 +186,73 @@ var _ = g.Describe("[sig-network][Feature:Router]", func() {
...
+					return false, err
+				}
+
+				if len(findGaugesWithLabels(updatedMetrics["haproxy_server_up"], serverLabels)) < 2 {
+					e2e.Logf("haproxy_server_up: %v", findGaugesWithLabels(updatedMetrics["haproxy_server_up"], serverLabels))
+					g.By("sending traffic to a weighted route")
+					return false, expectRouteStatusCodeRepeatedExec(ns, execPodName, fmt.Sprintf("http://%s", host), "weighted.metrics.example.com", http.StatusOK, times, proxyProtocol)
+				}
+
+				if findGaugesWithLabels(updatedMetrics["haproxy_backend_connections_total"], routeLabels)[0] < float64(times) {
```
Comment (sgreene570) [inline test/extended/router/metrics.go:222]: Are the other values in this slice not important? Would it be worth checking the entire slice?

**Comment by frobware** [comment]
No haproxy metrics failure in this run:

https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-gcp/8166

/retest

**Comment by frobware** [comment]
No haproxy metrics failure in this run:

https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-gcp/8178

/test e2e-gcp

**Comment by frobware** [comment]
No haproxy metrics failure in this run:

https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-gcp/8190

/test e2e-gcp

File: test/extended/router/metrics.go [hunk 3]
```diff
@@ -186,7 +186,73 @@ var _ = g.Describe("[sig-network][Feature:Router]", func() {
 			o.Expect(findGaugesWithLabels(metrics["haproxy_backend_up"], routeLabels)).To(o.Equal([]float64{1}))
 			o.Expect(findGaugesWithLabels(metrics["haproxy_server_bytes_in_total"], serverLabels)).To(o.ConsistOf(o.BeNumerically(">=", 0), o.BeNumerically(">=", 0)))
 			o.Expect(findGaugesWithLabels(metrics["haproxy_server_bytes_out_total"], serverLabels)).To(o.ConsistOf(o.BeNumerically(">=", 0), o.BeNumerically(">=", 0)))
-			o.Expect(findGaugesWithLabels(metrics["haproxy_server_max_sessions"], serverLabels)).To(o.ConsistOf(o.BeNumerically(">", 0), o.BeNumerically(">", 0)))
+
+			// haxproxy_server_max_sessions needs to be
```
Comment (smarterclayton) [inline test/extended/router/metrics.go:190]: this chunk probably belongs at the very end of the test, after we force delete the pod

**Comment by frobware** [comment]
This run https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-gcp/8217 did fail - this after moving the test to after the pod deletion per comment https://github.com/openshift/origin/pull/25045#discussion_r434211378.

/test e2e-gcp

**Comment by frobware** [comment]
> This run https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-gcp/8217 did fail - this after moving the test to after the pod deletion per comment [#25045 (comment)](https://github.com/openshift/origin/pull/25045#discussion_r434211378).

I need to look at this again to see why this failed. It may have been a one off but I'm suspicious.

File: test/extended/router/metrics.go (hunk 1 above)
Comment (frobware) [reply test/extended/router/metrics.go]: Not any more as the test got moved to the end of the suite and there we only have 1 endpoint now.

File: test/extended/router/metrics.go (hunk 3 above)
Comment (frobware) [reply test/extended/router/metrics.go:222]: I moved this and cancelled the hold. I can run it through a few more CI runs but this should now cut out a lot of the flakes.

**Comment by bparees** [comment]
this bug is killing us in CI, is there someone other than @smarterclayton who can review/approve?

**Comment by frobware** [comment]
> this bug is killing us in CI, is there someone other than @smarterclayton who can review/approve?

/unassign @smarterclayton 
/assign @knobunc

**Comment by frobware** [comment]
1 failure from https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-gcp/8307

    [sig-instrumentation][Late] Alerts shouldn't report any alerts in firing state apart from Watchdog and AlertmanagerReceiversNotConfigured [Suite:openshift/conformance/parallel]

/retest

**Comment by frobware** [comment]
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/25045/pull-ci-openshift-origin-master-e2e-cmd/5873

    [sig-cli][Feature:LegacyCommandTests][Disruptive][Serial] test-cmd: test/cmd/set-image.sh [Suite:openshift] 

/test e2e-cmd

**Pull request merged**